File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <unordered_map>
using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    // 장르: 재생횟수
    unordered_map<string,int> g;
    // 장르: 재생횟수, 음악번호
    unordered_map<string,vector<pair<int,int>>> c;
    
    for(int i=0;i<genres.size();i++){
        g[genres[i]]+=plays[i];
        c[genres[i]].push_back({plays[i],i});
    }
    // 1. 장르별 총 재생수를 기준으로 장르 정렬하기
    // 정렬된 장르를 넣을 수 있도록 
    vector<pair<string, int>> sorted_g(g.begin(), g.end());
    sort(sorted_g.begin(), sorted_g.end(), [](auto a, auto b) {
        return a.second > b.second;
    });
    
    // 2. 정렬된 장르 내에서 노래 고르기
    for(auto const&[genre,play]:sorted_g){
        //현재 장르
        // 재생횟수 순, 재생횟수 같으면 이름 순
        vector<pair<int, int>>& songs = c[genre];
        sort(songs.begin(),songs.end(),[](auto a, auto b){
            if (a.first==b.first){
                return a.second<b.second;
            }
            return a.first>b.first;
        });
            for(int i = 0; i < songs.size() && i < 2; i++) {
            answer.push_back(songs[i].second);
        }
    }
        
    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {

    class Song {
        int idx;
        int play;

        Song(int idx, int play) {
            this.idx = idx;
            this.play = play;
        }
    }

    public int[] solution(String[] genres, int[] plays) {

        HashMap<String, Integer> genrePlay = new HashMap<>();
        HashMap<String, ArrayList<Song>> genreSongs = new HashMap<>();

        for (int i = 0; i < genres.length; i++) {

            genrePlay.put(
                genres[i],
                genrePlay.getOrDefault(genres[i], 0) + plays[i]
            );

            genreSongs.putIfAbsent(
                genres[i],
                new ArrayList<>()
            );

            genreSongs.get(genres[i])
                      .add(new Song(i, plays[i]));
        }

        ArrayList<String> genreList =
            new ArrayList<>(genrePlay.keySet());

        genreList.sort(
            (a, b) -> genrePlay.get(b) - genrePlay.get(a)
        );

        ArrayList<Integer> result =
            new ArrayList<>();

        for (String genre : genreList) {

            ArrayList<Song> songs =
                genreSongs.get(genre);

            songs.sort((a, b) -> {
                if (a.play == b.play)
                    return a.idx - b.idx;

                return b.play - a.play;
            });

            for (int i = 0;
                 i < Math.min(2, songs.size());
                 i++) {

                result.add(songs.get(i).idx);
            }
        }

        return result.stream()
                     .mapToInt(i -> i)
                     .toArray();
    }
}